### PR TITLE
Make it easier to disable BruteForceLocker; #server-1345

### PR DESCRIPTION
**Pull request by woj-tek**

### CODE REVIEW

File: src/main/java/tigase/xmpp/impl/SaslAuth2.java [hunk 1]
```diff
@@ -484,7 +489,7 @@ private BareJID extractUserJid(final Exception e, XMPPResourceConnection session
 
 	private void saveIntoBruteForceLocker(final XMPPResourceConnection session, final Exception e) {
 		try {
-			if (bruteForceLocker.isEnabled(session)) {
+			if (bruteForceLocker != null && bruteForceLocker.isEnabled(session)) {
```
Comment (hantu85) [inline src/main/java/tigase/xmpp/impl/SaslAuth2.java:492]: As you have more than one place you are doing the same check (but in different files), I would think about adding method `bool isBruteForceLockerEnabled(XMPPResourceConnection)` and storing `null` check and `isEnabled()` inside this method.

File: src/main/java/tigase/xmpp/impl/SaslAuth.java [hunk 2]
```diff
@@ -211,7 +216,7 @@ public void process(final Packet packet, final XMPPResourceConnection session, f
 														  session.getDomain().getVhost().getDomain());
 						}
 
-						if (bruteForceLocker.isEnabled(session) &&
+						if (bruteForceLocker != null && bruteForceLocker.isEnabled(session) &&
```
Comment (hantu85) [inline src/main/java/tigase/xmpp/impl/SaslAuth.java:219]: The same as in comment for `SaslAuth2` about new method. Maybe it would be worth to add this new method to the new class `AbstractAuthWithBruteForceLockPreprocessor` and rebase those classes to use this new class?

Comment (hantu85) [reply src/main/java/tigase/xmpp/impl/SaslAuth.java:219]: Or maybe add BruteForceLocker to `AbstractAuthPreprocesor` and make that class a "wrapper" for using BruteForceLocker?

**Review by hantu85 — COMMENTED** [review]
I've made a few suggestions, but even without them, the code should work just fine.

File: src/main/java/tigase/xmpp/impl/SaslAuth.java (hunk 2 above)
Comment (woj-tek) [reply src/main/java/tigase/xmpp/impl/SaslAuth.java]: I moved `BruteForceLocker` checking to `AbstractAuthPreprocesor` as it made the most sense (it should be used by all auth plugins). While working on it, I also refactored most of the common parts of `SaslAuth` and `SaslAuth2` to `SaslAuthAbstract`.

File: src/main/java/tigase/xmpp/impl/AbstractAuthPreprocessor.java [hunk 3]
```diff
@@ -54,6 +60,14 @@
 			new ElementMatcher(Iq.IQ_QUERY_PATH, JabberIqAuth.ID, true)
 	};
 
+	protected boolean isBruteForceLockerEnabled(XMPPResourceConnection session) {
```
Comment (hantu85) [inline src/main/java/tigase/xmpp/impl/AbstractAuthPreprocessor.java:63]: I was thinking more about creating methods like `boolean isLoginAllowed(XMPPResourceConnection)` and `void reportInvalidLogin(session, clientIP, user_id)` and making BruteForceLockerBean field private as we could push all NPE checks in `AbstractAuthPreprocessor` class.

File: src/main/java/tigase/xmpp/impl/SaslAuthAbstract.java [hunk 4]
```diff
@@ -0,0 +1,211 @@
...
+abstract public class SaslAuthAbstract
+		extends AbstractAuthPreprocessor
+		implements XMPPProcessorIfc {
+
+	protected final static String SASL_SERVER_KEY = "SASL_SERVER_KEY";
+	protected final static String ALLOWED_SASL_MECHANISMS_KEY = "allowed-sasl-mechanisms";
+
+	private static final Logger log = Logger.getLogger(SaslAuthAbstract.class.getName());
+
+	public enum ElementType {
```
Comment (hantu85) [inline src/main/java/tigase/xmpp/impl/SaslAuthAbstract.java:53]: I'm not sure we should "merge" two sets of element types for 2 different SASL protocols into one enum (set/collection). I think any change in the name of elements in SASL2 (which is still possible) can result in a lot of errors as it may require updating in some places but not others. I would prefer to keep the protocol part separate from the logic part, so we could "merge" logic into one code, but keep 2 separate protocol implementations.

File: src/main/java/tigase/xmpp/impl/SaslAuthAbstract.java [hunk 5]
```diff
@@ -0,0 +1,211 @@
...
+	protected final Map<String, Object> props = new HashMap<>();
+	@Inject
+	protected TigaseSaslProvider saslProvider;
+
+	@Override
+	public int concurrentQueuesNo() {
+		return super.concurrentQueuesNo() * 4;
+	}
+
+	protected Element createReply(final SaslAuth2.ElementType type, final String cdata) {
```
Comment (hantu85) [inline src/main/java/tigase/xmpp/impl/SaslAuthAbstract.java:93]: As I'm not convinced on using "unified" ElementType, maybe we could just use `String` as a `type` here?

File: src/main/java/tigase/xmpp/impl/SaslAuth.java [hunk 6]
```diff
@@ -114,41 +83,15 @@ public void process(final Packet packet, final XMPPResourceConnection session, f
...
+				processSessionAlreadyAuthorized(packet, session, results);
 				return;
 			} else {
 				Element request = packet.getElement();
 				try {
 					SaslServer ss;
 
 					if ("auth" == request.getName()) {
-						final String mechanismName = request.getAttributeStaticStr("mechanism");
+						final String mechanismName = packet.getElement().getAttributeStaticStr("mechanism");
```
Comment (hantu85) [inline src/main/java/tigase/xmpp/impl/SaslAuth.java:94]: I think this change is unnecessary as we already have element instance `request`? so why this changes?  if it should be made, then maybe we should replace `request.getName()` in a line above as well?

File: src/main/java/tigase/xmpp/impl/SaslAuthAbstract.java (hunk 4 above)
Comment (woj-tek) [reply src/main/java/tigase/xmpp/impl/SaslAuthAbstract.java:53]: My reasoning was: it's still SASL so within same area. What's more - this is the list of possible Element types - I'm not sure that those should/would change, rather new would be added but addition wouldn't case issues with logic/how it's handled.

File: src/main/java/tigase/xmpp/impl/SaslAuthAbstract.java (hunk 5 above)
Comment (woj-tek) [reply src/main/java/tigase/xmpp/impl/SaslAuthAbstract.java:93]: `ElementType` should have been moved to SaslAbstract. As for the usage - I'm not convinced to use String here, if there is opposition then using separate enums for both Sasl and separate `createReply` would be better.

Still - this is a set of possible element types, which seems quite "share-able"

File: src/main/java/tigase/xmpp/impl/SaslAuth.java (hunk 6 above)
Comment (woj-tek) [reply src/main/java/tigase/xmpp/impl/SaslAuth.java:94]: I was just doing diff on `process()` method to see difference and it's a leftover.

File: src/main/java/tigase/xmpp/impl/SaslAuthAbstract.java (hunk 4 above)
Comment (hantu85) [reply src/main/java/tigase/xmpp/impl/SaslAuthAbstract.java:53]: There are 2 SASLs, different versions, so they can be different. I'm not keen on relaying on the fact "they are similar" currently.

File: src/main/java/tigase/xmpp/impl/SaslAuth.java (hunk 2 above)
Comment (hantu85) [reply src/main/java/tigase/xmpp/impl/SaslAuth.java]: OK

File: src/main/java/tigase/xmpp/impl/SaslAuth.java (hunk 6 above)
Comment (hantu85) [reply src/main/java/tigase/xmpp/impl/SaslAuth.java:94]: OK

File: src/main/java/tigase/xmpp/impl/SaslAuthAbstract.java (hunk 4 above)
Comment (woj-tek) [reply src/main/java/tigase/xmpp/impl/SaslAuthAbstract.java:53]: Well, there are 3 SASLs: old one, new one and proposed changes to new one...

File: src/main/java/tigase/xmpp/impl/SaslAuthAbstract.java (hunk 5 above)
Comment (woj-tek) [reply src/main/java/tigase/xmpp/impl/SaslAuthAbstract.java:93]: It's currently stuck? And if/when it's published and then if/when it has any changes we can decouple the enum if that's required? Maintaining a copy simply because something could happen seems inefficient.

Comment (hantu85) [reply src/main/java/tigase/xmpp/impl/SaslAuthAbstract.java:93]: Maintaining a copy of the list of names used by a particular XEP makes sense, as we can update enum exact implementation of the XEP in our code without impact on other parts of Tigase and also without making any API changes not connected to the XEP implementation that we are updating.

If you wish to leave it as it is, we can do that, but I'm not convinced that this is a way to go.

Comment (woj-tek) [reply src/main/java/tigase/xmpp/impl/SaslAuthAbstract.java]: Went with two separate enums.

File: src/main/java/tigase/xmpp/impl/AbstractAuthPreprocessor.java (hunk 3 above)
Comment (woj-tek) [reply src/main/java/tigase/xmpp/impl/AbstractAuthPreprocessor.java:67]: Moved all members operating on brute force locker instance to `AbstractAuthPreprocessor`

**Comment by woj-tek** [comment]
I also added rudimentary sasl2 unit test and... why it uses session manager directly for failure instead of using results queue?

**Comment by hantu85** [comment]
It is required for SASL2 to reply **after** binding (bind2) is completed if it takes place and during the binding process we are calling methods that are completing asynchronously, SASL2 sends results using SessionManager as using the results queue from a different thread is not an option.
